File: AtCoder/ABC/ABC233/C.js
'use strict';
// var input = require('fs').readFileSync('/dev/stdin', 'utf8').trim().split('\n');
var input = require('fs').readFileSync('../input.txt', 'utf8').trim().split('\n');

let cid = 0;
const getLine = (isStr = false) => { cid++; return isStr ? input[cid - 1].trim().split(' ') : input[cid - 1].split(' ').map(e => +e) };
const getLines = (n = 1, isStr = false) => { cid += n; return isStr ? input.slice(cid - n, cid).map(line => line.trim().split(' ')) : input.slice(cid - n, cid).map(line => line.split(' ').map(e => +e)) }
function xArray(v) { var a = arguments, l = a.length, r = 'Array(a[' + --l + ']).fill().map(x=>{return ' + v + ';})'; while (--l) r = 'Array(a[' + l + ']).fill().map(x=>' + r + ')'; return eval(r); }
var streams = []; function print(s) { streams.push(s); }

class Digit {
    constructor(array) {
        this.limit = array;
        this.array = xArray(0, array.length);
    }

    increment = () => {
        this.array[0] += 1;

        const inc = (index) => {
            try {
                this.array[index + 1] += Math.floor(this.array[index] / this.limit[index]);
                this.array[index] += (this.array[index]) % this.limit[index];
                if (index + 1 < this.limit.length)
                    inc(index + 1);
            } catch {
                return -1;
            }
        }
        return inc(0);
    }
}

function main() {
    const [N, X] = getLine();
    let length = [];
    let mat = [];
    for (let _ = 0; _ < N; _++) {
        const num = getLine();
        length.push(num[0])
        mat.push(num.slice(1));
    }

    // 全探索
    let digit = new Digit(length);
    while (digit.increment() != -1) {
        console.log(`digit.array`, digit.array)
    }
};

var myOut = main();
if (myOut !== undefined) console.log(String(myOut));
if (streams.length) console.log(streams.join('\n'));

File: AtCoder/ABC/ABC233/C.py
import io
import sys

# _INPUT = '''\
# 3 1000000000000000000
# 2 1000000000 1000000000
# 2 1000000000 1000000000
# 2 1000000000 1000000000

# '''
# sys.stdin = io.StringIO(_INPUT)

[N, X] = [int(_) for _ in input().split()]
mat = [[int(_) for _ in input().split()] for i in range(N)]

# 深さ優先探索で、全探索する


def dfs(depth, product):

    if depth == N-1:
        # print(product)
        return 1 if product == X else 0
    else:
        _ans = 0
        for idx, val in enumerate(mat[depth+1]):
            if idx == 0:
                continue
            _ans = _ans + dfs(depth+1, product*val)
        return _ans


ans = 0
for idx, val in enumerate(mat[0]):
    if idx == 0:
        continue
    ans = ans + dfs(0, val)
print(ans)
